### PR TITLE
Dual tor failure utilities and sample tests

### PR DESCRIPTION
**Pull request by vaibhavhd**
### Description of PR

Summary: Add common utilities for dual ToR failure
Fixes # (issue) https://github.com/Azure/sonic-mgmt/issues/2757

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [x] Test case(new/improvement)

#### What is the motivation for this PR?
ToR Failure utilities. This change includes fixtures to:
1. Reboot TOR
2. Blackhole traffic
3. Shutdown bgp sessions
4. Shutdown LinkProber (to be tested after LinkProber implementation is complete).

Some tests are added to verify functionality of these utilities.
Additionally Dual ToR failure testcase definitions are added. The implementation will be taken care in the future PRs.
#### How did you verify/test it?

Tested utilities using sample tests on a dualtor testbed with MUX simulator running in server.

### CODE REVIEW

File: tests/common/dualtor/tor_failure_utils.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import ipaddress
+import pytest
+import logging
+import time
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def shutdown_tor_bgp():
```
Comment (lolyu) [inline tests/common/dualtor/tor_failure_utils.py:18]: Will it be simpler to define those utilities as functions instead of fixtures and let the user import them to use?

Comment (theasianpianist) [reply tests/common/dualtor/tor_failure_utils.py:18]: Agree here. Could we make these methods so that the tests can decide when they want to call them? I think it would be easier to measure the traffic disruption this way, since we cannot directly control when a fixture is run.

File: tests/common/dualtor/tor_failure_utils.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        bgp_neighbors = duthost.get_bgp_neighbors()
+        up_bgp_neighbors = [ k.lower() for k, v in bgp_neighbors.items() if v["state"] == "established" ]
+        if shutdown_all and up_bgp_neighbors:
+            logger.info("Shutdown BGP sessions on {}".format(duthost.hostname))
+            duthost.shell("config bgp shutdown all")
+
+    yield shutdown_tor_bgp
+
+    time.sleep(1)
+    if torhost:
```
Comment (theasianpianist) [inline tests/common/dualtor/tor_failure_utils.py:34]: Why is this necessary? Are we expecting `duthost` to become undefined at some point during the test?

File: tests/common/dualtor/tor_failure_utils.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+def reboot_tor(localhost):
+    """
+    Reboot TOR
+    """
+    torhost = []
+    torhost_ip = []
+    def reboot_tor(duthost, reboot_type=REBOOT_TYPE_COLD):
+        torhost.append(duthost)
+        torhost_ip.append(duthost.setup()['ansible_facts']['ansible_eth0']['ipv4']['address'])
+        logger.info("Shutdown BGP sessions on {}".format(duthost.hostname))
```
Comment (theasianpianist) [inline tests/common/dualtor/tor_failure_utils.py:105]: Incorrect log message?

File: tests/common/dualtor/tor_failure_utils.py (hunk 1 above)
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py:18]: @lolyu I think Lawrence's comment is the right answer. In summary, I see two major benefits of keeping the fixtures:

1. Tests do not have to care for the cleanup/restoration. This also ensures if a test fails, the fixtures cleanup will still be called, instead of a try/catch/finally or cleanup in a test. Keeping tests appear simple.
2. Tests do not have to send every argument to the function. The outer fixture will keep session level fixtures available for inner logic. This will prevent putting args in test definitions and in the function calls. An example in this PR is a `reboot_tor` fixture which keeps localhost fixture available for inner function.

File: tests/common/dualtor/tor_failure_utils.py (hunk 3 above)
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py]: Addressed in the latest.

File: tests/dualtor/tor_failure/test_tor_component_failure.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+import pytest
```
Comment (theasianpianist) [inline tests/dualtor/tor_failure/test_tor_component_failure.py:1]: Should we save these files for a future PR? Not sure it makes sense to include them at the moment.

**Review by theasianpianist — CHANGES_REQUESTED** [review]

File: tests/common/dualtor/tor_failure_utils.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    def reboot_tor(duthost, reboot_type=REBOOT_TYPE_COLD):
+        torhost.append(duthost)
+        logger.info("Issuing reboot of type {} on {}".format(reboot_type, duthost.hostname))
+        reboot(duthost, localhost, reboot_type=reboot_type, wait_for_ssh=False)
+
+    yield reboot_tor
+    # TODO Add IO check capability
+
+    if torhost:
+        duthost = torhost[0]
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py:110]: Added.

File: tests/common/dualtor/tor_failure_utils.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        torhost.append(duthost)
+        if asic:
+            duthost.shell("ip route del 0.0.0.0/0")
+        elif kernel:
+            pass # TODO
+
+    yield tor_blackhole_traffic
+
+    if torhost:
+        duthost = torhost[0]
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py:77]: Added.

File: tests/common/dualtor/tor_failure_utils.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    def shutdown_tor_heartbeat(duthost):
+        # TODO - verify support after LinkProber submodule is ready
+        torhost.append(duthost)
+        duthost.shell("sudo systemctl stop mux")
+        duthost.shell("sudo systemctl disable mux")
+
+    yield shutdown_tor_heartbeat
+
+    if torhost:
+        duthost = torhost[0]
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py:55]: Added.

File: tests/common/dualtor/tor_failure_utils.py [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+@pytest.fixture
+def shutdown_tor_heartbeat():
+    """
+    Shutdown the LinkProber
+    """
+    torhost = []
+    def shutdown_tor_heartbeat(duthost):
+        # TODO - verify support after LinkProber submodule is ready
+        torhost.append(duthost)
+        duthost.shell("sudo systemctl stop mux")
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py]: Removed `sudo`.

File: tests/common/dualtor/tor_failure_utils.py [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+            if k == "Loopback0":
+                for ipstr in v.keys():
+                    ip = ipaddress.ip_interface(ipstr)
+                    if ip.version == 4:
+                        lo_ipv4 = ip
+                    elif ip.version == 6:
+                        lo_ipv6 = ip
+
+        duthost.shell("ip -4 route add 0.0.0.0/0 nexthop via {}".format(lo_ipv4.ip))
+        #duthost.shell("ip -6 route add ::/0 nexthop via {}".format(lo_ipv6.ip))
```
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py]: Removed the comment. Right now blackholing by only removing ipv4 default. This can be modified later if needed for ipv6.

File: tests/dualtor/tor_failure/test_tor_component_failure.py (hunk 4 above)
Comment (vaibhavhd) [reply tests/dualtor/tor_failure/test_tor_component_failure.py]: Good suggestion. Removed them for later.

File: tests/common/dualtor/tor_failure_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (theasianpianist) [reply tests/common/dualtor/tor_failure_utils.py:55]: I think this line (55) is no longer needed after adding the loop

File: tests/common/dualtor/tor_failure_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (theasianpianist) [reply tests/common/dualtor/tor_failure_utils.py:77]: No longer needed

File: tests/common/dualtor/tor_failure_utils.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py]: Ah, you are right. Removed.

File: tests/common/dualtor/tor_failure_utils.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vaibhavhd) [reply tests/common/dualtor/tor_failure_utils.py]: Removed.

**Review by theasianpianist — APPROVED** [review]
